library/src/main/java/com/datatorrent/lib/state/managed/BucketsFileSystem.java

```diff
@@ -0,0 +1,573 @@
+      try (DataInputStream dis = getInputStream(bucketId, META_FILE_NAME)) {
+        //Load meta info of all the time buckets of the bucket identified by bucketId.
+        loadBucketMetaFile(bucketId, dis);
+      }
+    } else {
+      return null;
+    }
+    return timeBucketsMeta.get(bucketId, timeBucketId);
+  }
+
+  /**
+   * Returns the meta information of all the time buckets in the bucket in descending order - latest to oldest.
+   *
+   * @param bucketId bucket id
+   * @return all the time buckets in order - latest to oldest
+   */
+  public TreeSet<ImmutableTimeBucketMeta> getAllTimeBuckets(long bucketId) throws IOException
+  {
+    synchronized (timeBucketsMeta) {
+      TreeSet<ImmutableTimeBucketMeta> immutableTimeBucketMetas = Sets.newTreeSet(
+          Collections.<ImmutableTimeBucketMeta>reverseOrder());
+
+      if (timeBucketsMeta.containsRow(bucketId)) {
+        for (Map.Entry<Long, MutableTimeBucketMeta> entry : timeBucketsMeta.row(bucketId).entrySet()) {
+          immutableTimeBucketMetas.add(entry.getValue().getImmutableTimeBucketMeta());
+        }
+        return immutableTimeBucketMetas;
+      }
+      if (exists(bucketId, META_FILE_NAME)) {
+        try (DataInputStream dis = getInputStream(bucketId, META_FILE_NAME)) {
+          //Load meta info of all the time buckets of the bucket identified by bucket id
+          loadBucketMetaFile(bucketId, dis);
+          for (Map.Entry<Long, MutableTimeBucketMeta> entry : timeBucketsMeta.row(bucketId).entrySet()) {
+            immutableTimeBucketMetas.add(entry.getValue().getImmutableTimeBucketMeta());
+          }
+          return immutableTimeBucketMetas;
+        }
+      }
+      return immutableTimeBucketMetas;
+    }
+  }
+
+  /**
+   * Loads the bucket meta-file.
+   *
+   * @param bucketId bucket id
+   * @param dis      data input stream
+   * @throws IOException
+   */
+  private void loadBucketMetaFile(long bucketId, DataInputStream dis) throws IOException
+  {
+    int numberOfEntries = dis.readInt();
+
+    for (int i = 0; i < numberOfEntries; i++) {
+      long timeBucketId = dis.readLong();
+      long dataSize = dis.readLong();
+      long lastTransferredWindow = dis.readLong();
+
+      MutableTimeBucketMeta tbm = new MutableTimeBucketMeta(bucketId, timeBucketId);
+
+      int sizeOfFirstKey = dis.readInt();
+      byte[] firstKeyBytes = new byte[sizeOfFirstKey];
+      dis.readFully(firstKeyBytes, 0, firstKeyBytes.length);
+      tbm.updateTimeBucketMeta(lastTransferredWindow, dataSize, new Slice(firstKeyBytes));
+
+      timeBucketsMeta.put(bucketId, timeBucketId, tbm);
+    }
+  }
+
+  /**
+   * Saves the updated bucket meta on disk.
+   *
+   * @param bucketId bucket id
+   * @throws IOException
+   */
+  void updateBucketMetaFile(long bucketId) throws IOException
+  {
+    Map<Long, MutableTimeBucketMeta> timeBuckets;
+    synchronized (timeBucketsMeta) {
+      timeBuckets = timeBucketsMeta.row(bucketId);
+    }
```

Comment:
synchronized block may have to cover most of this function because the map returned by the row() call is not a separate copy, so if another thread modifies the row while the loop below is iterating, there could be problems. The docs for the row() call say: "Changes to the returned map will update the underlying table, and vice versa."